### PR TITLE
Added keyword and location input

### PR DESCRIPTION
**Pull request by BernardIgiri**
Added keyword and location input using GeospatialDraw.

keyword-input

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     "eslint-plugin-graphql": "^3.1.0",
     "eslint-plugin-import": "^2.18.2",
     "eslint-plugin-react-hooks": "^2.1.2",
+    "fetch-mock": "^8.0.0",
```
Comment (djblue) [inline package.json]: All network interaction should happen though the graphql API. If you are using fetch directly, it's bypassing our network abstraction.

File: src/main/webapp/location/keyword.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        }}
+        getOptionLabel={(suggestion: Suggestion) => suggestion.name}
+        options={suggestions}
+        loading={loading}
+        inputValue={input}
+        onChange={(_e, value) => {
+          if (value) {
+            const { id: keywordId, name } = value
+            ;(async () => {
+              const response = await fetch(`./internal/geofeature?id=${id}`)
```
Comment (djblue) [inline src/main/webapp/location/keyword.tsx]: We should be accessing data via the graphql API

File: package.json (hunk 1 above)
Comment (BernardIgiri) [reply package.json]: Okay, I will have to figure out graphql on Monday. Thank you.

**Comment by BernardIgiri** [comment]
Alright, the GraphQL changes should be ready now.

File: src/main/webapp/intrigue-api/graphql.js [hunk 3]
```diff
@@ -7,7 +7,7 @@ import { mergeDeepOverwriteLists } from '../utils'
 const { BatchHttpLink } = require('apollo-link-batch-http')
 const { genSchema, toGraphqlName, fromGraphqlName } = require('./gen-schema')
 
-const ROOT = '/search/catalog/internal'
+const ROOT = 'https://localhost:8993/search/catalog/internal'
```
Comment (gjvera) [inline src/main/webapp/intrigue-api/graphql.js]: why is this change needed?

Comment (BernardIgiri) [reply src/main/webapp/intrigue-api/graphql.js]: I kept getting `only absolute urls are supported` errors until I did that in my tests of the GraphQL queries in `keyword.tsx` Not sure why but I just reverted this line.

File: src/main/webapp/location/geolocation.stories.js [hunk 4]
```diff
@@ -84,3 +90,103 @@ Object.keys(editors).forEach(key => {
...
+        [-77.3811, 42.7304],
+        [-77.3811, 42.5304],
+      ],
+    ],
+  },
+  properties: {},
+  id: 'Italy Valley Cemetery',
+}
+const extendedKeywordProps = {
+  getSuggestions: () => JSON.parse(JSON.stringify(keywordSuggestions)),
```
Comment (djblue) [inline src/main/webapp/location/geolocation.stories.js]: Why are we `JSON.stringify` then `JSON.parse` here?

File: src/main/webapp/intrigue-api/schema.graphql [hunk 5]
```diff
@@ -277,6 +277,18 @@ type FacetResult {
   value: String
 }
 
+type Suggestion {
+  id: String!
```
Comment (djblue) [inline src/main/webapp/intrigue-api/schema.graphql]:
```suggestion
  id: ID!
```

File: src/main/webapp/location/keyword.tsx [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    color,
+  } = properties as KeywordGeoProperties
+  const hasSelection = initKeyword && initKeywordId ? true : false
+  const [input, setInput] = React.useState<string>(initKeyword)
+  const [open, setOpen] = React.useState(false)
+  const [suggestions, setSuggestions] = React.useState<Suggestion[]>([])
+  const [{ keyword, keywordId }, setSelectedKeyword] = React.useState<
+    SelectedKeyword
+  >({})
+  React.useEffect(
```
Comment (djblue) [inline src/main/webapp/location/keyword.tsx]: I think this code could be simplified. I think we have too many effects happening.

File: src/main/webapp/location/keyword.tsx [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        query: FeatureQuery,
+        variables: {
+          id,
+        },
+      }),
+      setFeatureQueryState
+    ).then(r => r.geofeature)
+  }
+  const getSuggestions = async (q: string) => {
+    return runQuery(
```
Comment (djblue) [inline src/main/webapp/location/keyword.tsx]: I think you might be able to use `useQuery` here. You can leverage the skip option when you don't have enough information to run the query.

File: src/main/webapp/location/bbox.tsx [hunk 8]
```diff
@@ -1,23 +1,11 @@
...
-import { Filter, INTERSECTS, ANY_GEO, GEOMETRY } from './filter'
-import { geoToWKT } from './geo-to-wkt'
+import { coordinates as coordinateEditor } from 'geospatialdraw'
 import FormControl from '@material-ui/core/FormControl'
 import FormLabel from '@material-ui/core/FormLabel'
 import Point from './point'
 import Props from './geo-editor'
 import SpacedLinearContainer from '../spaced-linear-container'
 
-export const generateFilter = (geo: geometry.GeometryJSON): Filter => ({
```
Comment (djblue) [inline src/main/webapp/location/bbox.tsx]: Why is this being removed?

File: src/main/webapp/location/keyword.tsx [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+  } = coordinateEditor.geoToPolygonProps(value)
+  const {
+    keyword: initKeyword = '',
+    keywordId: initKeywordId,
+    color,
+  } = properties as KeywordGeoProperties
+  const hasSelection = initKeyword && initKeywordId ? true : false
+  const [input, setInput] = React.useState<string>(initKeyword)
+  const [open, setOpen] = React.useState(false)
+  const [suggestions, setSuggestions] = React.useState<Suggestion[]>([])
```
Comment (djblue) [inline src/main/webapp/location/keyword.tsx]: I think this state is already part of the apollo cache.

File: src/main/webapp/location/keyword.tsx (hunk 7 above)
Comment (BernardIgiri) [reply src/main/webapp/location/keyword.tsx]: Unfortunately useQuery cannot be converted to async functions which is how Keyword is bypassing Apollo in storybook.

File: src/main/webapp/location/bbox.tsx (hunk 8 above)
Comment (BernardIgiri) [reply src/main/webapp/location/bbox.tsx]: This is no longer needed, filters can be generated directly from GeometryJSON.

File: src/main/webapp/location/geolocation.stories.js (hunk 4 above)
Comment (BernardIgiri) [reply src/main/webapp/location/geolocation.stories.js]: To get a deep clone that isn't going to modify the original value and avoid any chances of side effects in the Storybook.

File: src/main/webapp/location/keyword.tsx (hunk 7 above)
Comment (Bdthomson) [reply src/main/webapp/location/keyword.tsx]: Why can't useQuery be used with async functions? Could you use `useLazyQuery` instead?

Comment (BernardIgiri) [reply src/main/webapp/location/keyword.tsx]: I'm going to refactor the async functions out. I'll be able to use `useLazyQuery` after that.

File: src/main/webapp/location/geolocation.stories.js (hunk 4 above)
Comment (djblue) [reply src/main/webapp/location/geolocation.stories.js]: Why is the suggestion at risk of being modified though?

File: src/main/webapp/location/keyword.tsx (hunk 7 above)
Comment (djblue) [reply src/main/webapp/location/keyword.tsx]: I think the keyword component shouldn't be async. It should accept props that can change over time.

File: src/main/webapp/location/bbox.tsx (hunk 8 above)
Comment (djblue) [reply src/main/webapp/location/bbox.tsx]: So the components speak geojson directly? :+1:

File: src/main/webapp/location/location.tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    shape: shapes.POLYGON,
+  },
+}
+
+const Location: React.SFC<Props> = ({
+  value,
+  onChange,
+  editorProps = {} as EditorPropsMap,
+}) => {
+  const [locationType, setLocationType] = React.useState<LocationType>(
```
Comment (samuelechu) [inline src/main/webapp/location/location.tsx]: If a component is taking in an onChange method, it shouldn't be keeping track of the state its calling the onChange method with. This data should be passed down as props.

File: src/main/webapp/location/keyword/container.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  value,
+  onChange,
+  minimumInputLength = MINIMUM_INPUT_LENGTH,
+  ...rest
+}) => {
+  const {
+    keyword = '',
+    keywordId = '',
+  } = value.properties as KeywordGeoProperties
+  const [selectedId, setSelectedId] = React.useState('')
```
Comment (samuelechu) [inline src/main/webapp/location/keyword/container.tsx]: I think you can move this state higher up in the component tree

Comment (samuelechu) [reply src/main/webapp/location/keyword/container.tsx]: both input and selectedId

Comment (BernardIgiri) [reply src/main/webapp/location/keyword/container.tsx]: There is no higher up component. The keyword container is just a GeoJSON in GeoJSON out interface. Internal details are not exposed this. This matches the interface provided by the other location input fields allowing for polymorphism.
